### PR TITLE
`PPDataset`: implement `BucketOrdering`

### PR DESCRIPTION
**Pull request by NeoLegends**
Let's see how well this works for the models.

### CODE REVIEW

**Comment by albertz** [comment]
Please check failing tests.
```
  returnn/datasets/postprocessing.py:301: WEAK WARNING PyMissingOrEmptyDocstringInspection: Missing docstring
  returnn/datasets/postprocessing.py:329: WEAK WARNING Pep8CodeStyle: E741: ambiguous variable name 'l'
  returnn/datasets/postprocessing.py:359: WEAK WARNING PyShadowingBuiltinsInspection: Shadows built-in name 'len'
```

File: returnn/datasets/postprocessing.py [hunk 1]
```diff
@@ -296,7 +298,70 @@ def _make_tensor_template_from_input(self, data_key: str) -> Tensor:
         return Tensor(data_key, dims=dims, dtype=dtype, sparse_dim=sparse_dim)
 
 
-class LaplaceOrdering(Callable[[Iterator[TensorDict]], Iterator[TensorDict]]):
+class SeqOrdering:
```
Comment (albertz) [inline returnn/datasets/postprocessing.py:301]: I don't see that this is such generic base class that I would call/use it that way. Maybe better:
```suggestion
class _GetSeqLenByLengthKey:
```

**Comment by albertz** [comment]
I don't really understand the purpose of this. This is not at the batch level here but still at the sequence level? So the batching comes after that and is independent? So this does not make sense? Bucketing is usually a way to do the batching. I.e. this is an alternative to our current `BatchingIterDataPipe`. Or we introduce sth like `batch_idx` and follow some logic as discussed in #1506.

File: returnn/datasets/postprocessing.py [hunk 2]
```diff
@@ -296,7 +298,73 @@ def _make_tensor_template_from_input(self, data_key: str) -> Tensor:
...
+    Iterator compatible with :class:`PostprocessingDataset`'s ``map_seq_stream`` applying
+    an ordering that creates batches consisting of segments falling into a few distinct
+    length buckets.
+
+    To be composed with any custom data postprocessing logic via :class:`Sequential`.
+
+    See also :class:`LaplaceOrdering`.
+    """
+
+    def __init__(self, seq_lens: Sequence[int], length_key: str = "data"):
```
Comment (albertz) [inline returnn/datasets/postprocessing.py:326]: For every max_seq_len, you should define the max_num_seqs. The option here should be more like:
```python
bucket_sizes: List[Tuple[int, int]]
```
representing the (max_seq_len,max_num_seqs) pair.

**Comment by NeoLegends** [comment]
The idea is that this yields the segments so they group up to batches in the desired way later in the pipeline w/o explicitly doing the batching. This is kind of a hack and I agree, this could also be done on the level of `BatchingIterDataPipe` or via `batch_idx`. I think I'll wait until I have results for the models and then we can decide whether to integrate this functionality deeper into the pipeline (which would make sense) or if we drop this PR completely.

**Comment by NeoLegends** [comment]
So maybe let's not merge this yet.

**Pull request closed without merging**